### PR TITLE
Validate: report more detailed errors

### LINKED ISSUE #2349: validate.cpp: display line numbers
Display violating line numbers, for some maximum line count.
Save lines numbers in arrays, print them along with an error.

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolves #2349

### CODE REVIEW

File: tools/validate/validate.cpp [hunk 1]
```diff
@@ -98,25 +92,48 @@ void scan_file(const filesystem::path& filepath, const TabPolicy tab_policy, vec
...
                 // https://en.wikipedia.org/wiki/ASCII#Printable_characters
                 ++disallowed_characters;
-                constexpr size_t MaxErrorsForDisallowedCharacters = 10;
-                if (disallowed_characters <= MaxErrorsForDisallowedCharacters) {
-                    fwprintf(stderr, L"Validation failed: %ls contains disallowed character 0x%02X.\n",
-                        filepath.c_str(), static_cast<unsigned int>(ch));
+
+                if (disallowed_characters <= MaxErrorsForErrorLinesReported) {
+                    fwprintf(stderr, L"Validation failed: %ls contains disallowed character 0x%02X at line %u.\n",
+                        filepath.c_str(), static_cast<unsigned int>(ch), static_cast<unsigned int>(line));
```
Comment (strega-nil-ms) [inline tools/validate/validate.cpp:98]:
```suggestion
                    fwprintf(stderr, L"Validation failed: %ls contains disallowed character 0x%02X at line %zu.\n",
                        filepath.c_str(), static_cast<unsigned int>(ch), line);
```

File: tools/validate/validate.cpp [hunk 2]
```diff
@@ -68,26 +68,20 @@ void scan_file(const filesystem::path& filepath, const TabPolicy tab_policy, vec
     size_t tab_characters            = 0;
     size_t trailing_whitespace_lines = 0;
 
+    constexpr size_t MaxErrorsForErrorLinesReported = 10; // per each error type we report
+
+    array<size_t, MaxErrorsForErrorLinesReported> overlength_line_occurecnes;
```
Comment (strega-nil-ms) [inline tools/validate/validate.cpp:73]:
```suggestion
    array<size_t, MaxErrorsForErrorLinesReported> lines_over_length_restriction{};
```

I also really don't understand why we're putting this here as opposed to printing as they come up...

File: tools/validate/validate.cpp [hunk 3]
```diff
@@ -68,26 +68,20 @@ void scan_file(const filesystem::path& filepath, const TabPolicy tab_policy, vec
     size_t tab_characters            = 0;
     size_t trailing_whitespace_lines = 0;
 
+    constexpr size_t MaxErrorsForErrorLinesReported = 10; // per each error type we report
```
Comment (strega-nil-ms) [inline tools/validate/validate.cpp:71]:
```suggestion
    constexpr size_t MaxErrorsReportedPerErrorType = 10; // per each error type we report
```

File: tools/validate/validate.cpp [hunk 4]
```diff
@@ -98,25 +92,48 @@ void scan_file(const filesystem::path& filepath, const TabPolicy tab_policy, vec
...
+                    fwprintf(stderr, L"Validation failed: %ls contains disallowed character 0x%02X at line %u.\n",
+                        filepath.c_str(), static_cast<unsigned int>(ch), static_cast<unsigned int>(line));
                 }
             }
 
             if (ch == CR || ch == LF) {
                 if (prev == ' ' || prev == '\t') {
                     ++trailing_whitespace_lines;
+
+                    if (overlength_lines <= MaxErrorsForErrorLinesReported) {
```
Comment (strega-nil-ms) [inline tools/validate/validate.cpp:106]: why does this use the `overlength_lines` variable?

File: tools/validate/validate.cpp [hunk 5]
```diff
@@ -98,25 +92,48 @@ void scan_file(const filesystem::path& filepath, const TabPolicy tab_policy, vec
...
+                } else {
+                    has_cr = true;
+                }
+                ++line;
+            } else {
+                if (ch == LF) {
+                    has_lf = true;
+                    ++line;
+                }
+            }
```
Comment (strega-nil-ms) [inline tools/validate/validate.cpp:135]: I know this is existing but

```suggestion
            } else if (ch == LF) {
                has_lf = true;
                ++line;
            }
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
I like the idea, but I've got some serious questions. Also:

- it seems unfortunate to not print the line number for all of the validation errors
- we should probably use the standard way of printing file:line:column (or at least file:line); something like:
  ```
  foo.cpp:10: error: non-printable character \xFF used
  ```

File: tools/validate/validate.cpp (hunk 2 above)
Comment (AlexGuteniev) [reply tools/validate/validate.cpp:73]: We first check for overlength lines, then if found, check for extension:
https://github.com/microsoft/STL/blob/314f65f831f827dbfb05ebcc0f08c496f5ddf64f/tools/validate/validate.cpp#L166-L185

Then finally print.

So in best case we avoid checking extension (a bit expensive operation), still with overlength lines we avoid printing for mismatched extension (must avoid).

Printing as we go with still avoiding unnecessary extension check will make us checking on the first overlength line. Which is probably fine, but makes this change more complicated than it was originally intended.

**Comment by AlexGuteniev** [comment]
> * we should probably use the standard way of printing file:line:column (or at least file:line); something like:

Good idea.
But isn't the standard for Visual Studio `file(line)` ?

A similar question about output format is asked in my PR #2502 , where it matters much more, as it would be end user visible.

**Comment by StephanTLavavej** [comment]
MSVC indeed emits `file(line)` and `file(line,column)`:

```
C:\Temp>type meow.cpp
class Cat {
    void meow() {}
};

int main() {
    Cat peppermint;
    peppermint.meow();
}

C:\Temp>cl /EHsc /nologo /W4 meow.cpp
meow.cpp
meow.cpp(7): error C2248: 'Cat::meow': cannot access private member declared in class 'Cat'
meow.cpp(2): note: see declaration of 'Cat::meow'
meow.cpp(1): note: see declaration of 'Cat'

C:\Temp>cl /EHsc /nologo /W4 /diagnostics:caret meow.cpp
meow.cpp
meow.cpp(7,21): error C2248: 'Cat::meow': cannot access private member declared in class 'Cat'
    peppermint.meow();
                    ^
meow.cpp(2): note: see declaration of 'Cat::meow'
    void meow() {}
meow.cpp(1): note: see declaration of 'Cat'
class Cat {
```

Seems like a reasonable precedent to imitate.

**Comment by StephanTLavavej** [comment]
There are merge conflicts that need to be resolved - and @strega-nil-ms points out that validate can now print VSO logging commands for this detailed error info.

**Comment by AlexGuteniev** [comment]
I'm restarting implementing, it would be easier than merging

**Comment by StephanTLavavej** [comment]
@AlexGuteniev Are you planning to continue work on this, or should we close it? @strega-nil-ms had significantly updated `validate.cpp`; after looking at the current state, her changes have the ability to send line numbers to the VSO log machinery but currently don't take advantage of that, so reporting more details would still be possible.

**Comment by AlexGuteniev** [comment]
I want to proceed, based on those @strega-nil-ms changes, but din't expect it soon.We can close the PR and I'll open a new one when I'm ready

**Comment by StephanTLavavej** [comment]
Ok, sounds good!

**Pull request closed without merging**